src/sinks/prometheus/exporter.rs

```diff
@@ -276,39 +278,60 @@ impl PrometheusExporter {
 impl StreamSink for PrometheusExporter {
     async fn run(&mut self, mut input: BoxStream<'_, Event>) -> Result<(), ()> {
         self.start_server_if_needed();
-        while let Some(event) = input.next().await {
-            let item = event.into_metric();
-            let mut metrics = self.metrics.write().unwrap();
-
-            match item.kind {
-                MetricKind::Incremental => {
-                    let new = MetricEntry(item.to_absolute());
-                    if let Some(MetricEntry(mut existing)) = metrics.take(&new) {
-                        if item.value.is_set() {
-                            // sets need to be expired from time to time
-                            // because otherwise they could grow infinitelly
-                            let now = Utc::now().timestamp();
-                            let interval = now - *self.last_flush_timestamp.read().unwrap();
-                            if interval > self.config.flush_period_secs as i64 {
-                                *self.last_flush_timestamp.write().unwrap() = now;
-                                existing.reset();
+
+        future::poll_fn(|cx| -> Poll<Result<(), ()>> {
+            let mut received = 0;
+            let mut metrics = None;
+
+            let result = loop {
+                let polled = Pin::new(&mut input).poll_next(cx);
+                match polled {
+                    Poll::Ready(Some(event)) => {
+                        received += 1;
+
+                        let metrics = metrics
+                            .get_or_insert_with(|| self.metrics.write().expect("poisoned lock"));
+
+                        let item = event.into_metric();
+                        match item.kind {
+                            MetricKind::Incremental => {
+                                let new = MetricEntry(item.to_absolute());
+                                if let Some(MetricEntry(mut existing)) = metrics.take(&new) {
+                                    if item.value.is_set() {
+                                        // sets need to be expired from time to time
+                                        // because otherwise they could grow infinitelly
+                                        let now = Utc::now().timestamp();
+                                        let interval =
+                                            now - *self.last_flush_timestamp.read().unwrap();
+                                        if interval > self.config.flush_period_secs as i64 {
+                                            *self.last_flush_timestamp.write().unwrap() = now;
+                                            existing.reset();
+                                        }
+                                    }
```

Comment:
@bruceg is this correct logic? Does it look like only one metric will be reset each flush period while we need to reset all the metrics, or I miss something?

Reply:
I agree with your reading, it looks broken, and would be a pre-existing bug with this sink. To be clear, only one *set* metric will be reset each flush period. This would be worth a test to prove out.